File: app/src/main/java/com/prisyazhnuy/newsapp/data/NewsRepository.java
package com.prisyazhnuy.newsapp.data;

import com.prisyazhnuy.newsapp.data.pojo.NewsResponse;
import com.prisyazhnuy.newsapp.data.remote.NewsApiService;

import io.reactivex.Observable;

/**
 * max.pr on 02.03.2018.
 */

public class NewsRepository implements NewsModel {

    private final NewsApiService mApiService;
    private static NewsRepository sInstance;

    private NewsRepository(NewsApiService apiService) {
        this.mApiService = apiService;
    }

    public static NewsModel getInstance(NewsApiService apiService) {
        if (sInstance == null) {
            sInstance = new NewsRepository(apiService);
        }
        return sInstance;
    }

    @Override
    public Observable<NewsResponse> getAllNews(int pageSize, int page) {
        return mApiService.getAllNews(pageSize, page);
    }

    @Override
    public Observable<NewsResponse> getNewsByFilter(String source, String from, String to, String sortBy, int pageSize, int page) {
        return mApiService.getNewsByFilter(source, from, to, sortBy, pageSize, page);
    }
}

File: app/src/main/java/com/prisyazhnuy/newsapp/dataKotlin/NewsRepository.kt
package com.prisyazhnuy.newsapp.dataKotlin

import android.util.Log
import com.prisyazhnuy.newsapp.dataKotlin.database.DataBaseCreator
import com.prisyazhnuy.newsapp.dataKotlin.database.repos.NewsDatabaseRepoImpl
import com.prisyazhnuy.newsapp.dataKotlin.models.News
import com.prisyazhnuy.newsapp.dataKotlin.network.RestClientBuilder
import io.reactivex.Flowable

/**
 * max.pr on 18.03.2018.
 */
interface NewsRepository {
    fun getBreakingNews(source: String, sortBy: String?, from: String?, to: String?, page: Int = 1): Flowable<List<News>>

    fun getFavouriteNews(): Flowable<List<News>>

    fun addToFavourites(news: News)

    fun removeFromFavourites(news: News)
}


object NewsRepositoryImpl : NewsRepository {

    private val newsNetworkRepo = RestClientBuilder.getNewsNetworkRepo()
    private val newsDatabaseRepo = NewsDatabaseRepoImpl(DataBaseCreator.dataBase.newsDao())

    override fun getBreakingNews(source: String, sortBy: String?, from: String?, to: String?, page: Int): Flowable<List<News>> {
        return newsNetworkRepo.getNewsList(source, from, to, sortBy, 25, page)
    }

    override fun getFavouriteNews(): Flowable<List<News>> {
        return newsDatabaseRepo.getNewsList()
                .doOnNext { data -> Log.d("NewsRepository", "data: $data") }
                .compose {
                    it.map {
                        arrayListOf<News>().apply {
                            it.forEach {
                                add(News(it.id, it.title, it.description, it.url, it.urlToImage))
                            }
                        }.toList()
                    }
                }
    }

    override fun addToFavourites(news: News) {
        newsDatabaseRepo.saveNews(news)
    }

    override fun removeFromFavourites(news: News) {
        newsDatabaseRepo.deleteNews(news)
    }
}